include/revng/Model/ABI.h

```diff
@@ -70,6 +70,10 @@ type: enum
       differences. The documentation can be found
       \sa https://docs.microsoft.com/en-us/cpp/cpp/cdecl
 
+  - name: Microsoft_x86_cdecl_gcc
+    doc: |
+      Turns out gcc's version of `cdecl` is subtly different from the original
```

Comment:
Wait, how is mingw even related? This is specifically the result of adding `__attribute__((ms_abi,cdecl))` (and friends) to the function prototype.

Reply:
> To be frank, this was a non-goal that has been pulled in without agreement.

This wasn't pulled in. This is the remainder from the previous PR where these were the only microsoft ABIs we _were_ testing.

2. Sadly, the real ones are too different, so it's impossible to keep them with the same name - hence "_gcc" suffix.

Reply:
> how much does this slow down the testing pipeline

3. There are 3 of these. Out of 16 total (so 3/13 (20%)) slowdown _worst_ case (because x86 gcc ones are generally pretty fast, compared to stuff like ARM or, worse, msvc).

Reply:
> This wasn't pulled in. This is the remainder from the previous PR where these were the only microsoft ABIs we were testing.

Right, but I thought we'd drop the old stuff.

In my mind, the GCC toolchain for Windows is MinGW, but yeah, technically it's a GCC feature, even if probably the mingw guys implemented it.

Anyway, let's keep it and enable testing.

Btw, can you make an example of a prototype that would not be compatible between GCC and MSVC and why?

Reply:
> MS breaks ABI at every major release

I don't think it's the same _ABI_, that's talked about.

This means the binary-interface side of things. Stuff like gcc/clang's `std::regex` or `std::function` stuck in an inferior state because changing some fields would mean binaries compiled with an older version is no longer compatible.

We, on the other hand, only think about the "basic" ABI level, as in which registers are used, how is stack space allocated for arguments, how they aligned, etc. I can't imagine this changing much, if ever.

> I wonder if GCC is maybe implementing and older MSVC ABI version

I severely doubt it. GCC's implementations _looks_ very similar to those of `regparm`s, and in general, they are very close to natural GCC ABIs. I'd imagine that's because the implementation is very crude, but still works for basic use cases (like when you need to call a function or two passing a couple of ints or something - without anything non-trivial), so that's why it's kept. But yeah, that's just my speculation.

Reply:
> I don't think it's the same ABI, that's talked about.

I think I agree... but it'd be nice to make a quick check, if we can, with some other MSVCs that we have available.
Which version(s) of MSVC are we using right now?

Also:

> Can you make an example of a prototype that would not be compatible between GCC and MSVC and why?

Reply:
> Which version(s) of MSVC are we using right now?

```
toolchain/win32-vc19/vc
```

> Can you make an example of a prototype that would not be compatible between GCC and MSVC and why?

for `cdecl`/`stdcall` the easiest one is return value:
`msvc` returns small structs (<=8 bytes) in registers (rdx+rax) which `gcc` always uses SPTAR (this one would be believable if it was a recent-ish change and old `msvc` also used SPTAR)

for `fastcall` (also `thiscall`, but that's way to broken), `gcc` uses `rcx` for for SPTAR, which `msvc` - stack. (this one I would not believe to be version related - there's no reason to "degrade" the SPTAR performance by suddenly starting to use stack).

Reply:
OK, we also have `vc{12,13,16}`, correct?
Do you think you can give a quick shot at some of them to see what happens?

I don't want an in-depth analysis, but at least understand if the situations is 1) they are all identical, 2) they 16 and 19 are almost the same, earlier are different, 3) they all are wildly different.

Maybe there's some reason due to which this is not easy to test, let me know.

Reply:
`vc12` also uses registers:
```cpp
// source
struct two_words {
  int a, b;
};

struct two_words global;
struct two_words whatever(void) { return global; }

// vc12
  401000: a1 28 54 40 00               	movl	0x405428, %eax
  401005: 8b 15 2c 54 40 00            	movl	0x40542c, %edx
// vc19
  401010: a1 00 53 41 00               	movl	0x415300, %eax
  401015: 8b 15 04 53 41 00            	movl	0x415304, %edx
```

Reply:
`gcc` 9.2.0 from our out toolchain with `__attribute((ms_abi, cdecl))`:
```cpp
010001a0 <whatever>:
 10001a0: 8b 44 24 04                  	movl	0x4(%esp), %eax
 10001a4: 8b 15 78 31 00 01            	movl	0x1003178, %edx
 10001aa: 8b 0d 7c 31 00 01            	movl	0x100317c, %ecx
 10001b0: 89 10                        	movl	%edx, (%eax)
 10001b2: 89 48 04                     	movl	%ecx, 0x4(%eax)
 10001b5: c3                           	retl
```

Reply:
I was not just talking about this specific situation, I was curious to see if the test suite for MSVC ABIs with binaries produced by older MSVC versions would still work.

But maybe it's better to back log this.

Reply:
I don't remember why specifically (or even if it's still in there - switching to event handlers turned everything upside down), but we explicitly depended on c11. (and `vc16` shipped with `visual studio 2010`, which cannot have c11 support by definition).

Reply:
OK, please backlog to check if we can relax constraints on the compiler and test older MSVC, not super urgent.